File: leetcode/Python/construct-binary-search-tree-from-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        def _traverse(lower, upper, cur):
            if idx[0]==len(preorder):
                return
            if lower<preorder[idx[0]]<upper and preorder[idx[0]]<cur.val:
                cur.left = TreeNode(preorder[idx[0]])
                idx[0]+=1
                _traverse(lower, cur.val, cur.left)
            if idx[0]<len(preorder) and lower<preorder[idx[0]]<upper and preorder[idx[0]]>cur.val:
                cur.right = TreeNode(preorder[idx[0]])
                idx[0]+=1
                _traverse(cur.val, upper, cur.right)
        sol, idx = TreeNode(preorder[0]), [1]
        _traverse(0,10000,sol)
        return sol

File: leetcode/C++/construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int idx;
    void traverse(vector<int>& preorder, int upper, int lower, TreeNode *cur){
        if(idx==preorder.size()) return;
        if(preorder[idx]<upper && preorder[idx]>lower && preorder[idx]<cur->val){
            cur->left=new TreeNode(preorder[idx++]);
            traverse(preorder, cur->val, lower, cur->left);
        }
        if(idx<preorder.size() && preorder[idx]<upper && preorder[idx]>lower && preorder[idx]>cur->val){
            cur->right=new TreeNode(preorder[idx++]);
            traverse(preorder, upper, cur->val, cur->right);
        }
    }
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        TreeNode * sol = new TreeNode(preorder[idx++]);
        traverse(preorder, 10000, 0 , sol);
        return sol;
    }
};


/*
o(n) 솔루션 찾기 어려웠음.
*/